src/models/layout/rule.rs

```diff
@@ -0,0 +1,629 @@
+        assert_eq!(json, json_serialize, "{:#?} != {:#?}", json, json_serialize);
+    }
+
+    #[test]
+    fn serialize_other() {
+        let rule = ArtifactRuleBuilder::new()
+            .set_type("CREATE")
+            .pattern("./artifact")
+            .build()
+            .unwrap();
+
+        let json = json!(["CREATE", "./artifact"]);
+        let json_serialize = serde_json::to_value(&rule).unwrap();
+        assert_eq!(json, json_serialize, "{:#?} != {:#?}", json, json_serialize);
+    }
+
+    #[test]
+    fn deserialize_full() {
+        let json = r#"[
+            "MATCH",
+            "foo.tar.gz",
+            "IN",
+            "./src",
+            "WITH",
+            "PRODUCTS",
+            "IN",
+            "./dst",
+            "FROM",
+            "package"
+        ]"#;
+        let rule = ArtifactRuleBuilder::new()
+            .set_type("MATCH")
+            .pattern("foo.tar.gz")
+            .source_path_prefix("./src")
+            .products()
+            .destination_path_prefix("./dst")
+            .step("package")
+            .build()
+            .unwrap();
+
+        let rule_parsed: ArtifactRule = serde_json::from_str(json).unwrap();
+        assert_eq!(rule, rule_parsed);
+    }
+
+    #[test]
+    fn deserialize_match_without_source() {
+        let json = r#"[
+            "MATCH",
+            "foo.tar.gz",
+            "WITH",
+            "PRODUCTS",
+            "IN",
+            "./dst",
+            "FROM",
+            "package"
+        ]"#;
+        let rule = ArtifactRuleBuilder::new()
+            .set_type("MATCH")
+            .pattern("foo.tar.gz")
+            .products()
+            .destination_path_prefix("./dst")
+            .step("package")
+            .build()
+            .unwrap();
+
+        let rule_parsed: ArtifactRule = serde_json::from_str(json).unwrap();
+        assert_eq!(rule, rule_parsed);
+    }
+
+    #[test]
+    fn deserialize_without_dst() {
+        let json = r#"[
+            "MATCH",
+            "foo.tar.gz",
+            "IN",
+            "./src",
+            "WITH",
+            "PRODUCTS",
+            "FROM",
+            "package"
+        ]"#;
+        let rule = ArtifactRuleBuilder::new()
+            .set_type("MATCH")
+            .pattern("foo.tar.gz")
+            .source_path_prefix("./src")
+            .products()
+            .step("package")
+            .build()
+            .unwrap();
+
+        let rule_parsed: ArtifactRule = serde_json::from_str(json).unwrap();
+        assert_eq!(rule, rule_parsed);
+    }
+
+    #[test]
+    fn deserialize_other() {
+        let json = r#"[
+            "DELETE",
+            "foo.pyc"
+        ]"#;
+        let rule = ArtifactRuleBuilder::new()
+            .set_type("DELETE")
+            .pattern("foo.pyc")
+            .build()
+            .unwrap();
+
+        let rule_parsed: ArtifactRule = serde_json::from_str(json).unwrap();
+        assert_eq!(rule, rule_parsed);
+    }
+}
```

Comment:
Broadly speaking, I like this "factory" for match rules. I think we should also at least ticketize a docs.rs pass for this or so.

Reply:
Good suggestion!  I've add some docs for it, but unfortunately rust clippy test did not pass. I didn't change anything about the error file `crypto.rs`, very weird.

Reply:
Yeah, I've seen this happen before. We may refactor `crypto.rs` during gsoc so I'm not exactly worried